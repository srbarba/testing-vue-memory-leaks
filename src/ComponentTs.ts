import Vue from 'vue'
import { mapGetters } from 'vuex'

export default Vue.extend({
  computed: {
    ...mapGetters(['userName']),
  },
  render(h) {
    return h('div', `Hello, ${this.userName}`)
  },
})
