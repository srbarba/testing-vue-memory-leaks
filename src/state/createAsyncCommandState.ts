import Vue from 'vue'
import deepmerge from 'deepmerge'

export const STATUS = {
  IDLE: 'IDLE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
}

const initialState = <U>(initialData: U) =>
  Vue.observable({
    status: STATUS.IDLE,
    data: deepmerge<U>({}, initialData),
    error: undefined as Error | undefined,
  })

export function createAsyncCommandState<U>(initialData = {} as U) {
  const state = initialState(initialData)

  return {
    getState() {
      return state
    },
    getData() {
      return state.data
    },
    getError() {
      return state.error
    },
    getStatus() {
      return state.status
    },
    isLoading() {
      return state.status === STATUS.LOADING
    },
    isError() {
      return state.status === STATUS.ERROR
    },
    isSuccess() {
      return state.status === STATUS.SUCCESS
    },
    setLoading() {
      state.status = STATUS.LOADING
    },
    setSuccess() {
      state.status = STATUS.SUCCESS
    },
    setData<U>(data: U) {
      state.status = STATUS.IDLE
      state.error = undefined
      Vue.set(state, 'data', Object.assign(state.data, data))
    },
    setError(appError?: Error) {
      state.status = STATUS.ERROR
      state.error = appError
    },
    reset() {
      Object.assign(state, initialState(initialData))
    },
  }
}
