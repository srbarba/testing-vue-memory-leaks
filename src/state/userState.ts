import type { User } from '../types/User'
import { createAsyncCommandState } from './createAsyncCommandState'

const initialState = {
  id: 0,
  name: '',
} as User

const commandState = createAsyncCommandState(initialState)
export const reset = commandState.reset

export const computedUserName = {
  set(name: User['name']) {
    commandState.setData({ name })
  },
  get() {
    return commandState.getData().name
  },
}

export const computedId = {
  set(id: User['id']) {
    commandState.setData({ id })
  },
  get() {
    return commandState.getData().id
  },
}
