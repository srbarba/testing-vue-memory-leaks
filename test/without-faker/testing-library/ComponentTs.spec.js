import { renderWithVuetify } from '../../renderWithVuetify'
import ComponentTs from '../../../src/ComponentTs'
import * as userState from '../../../src/state/userState'
import { UserMother } from '../../UserMotherWithoutFaker'

describe('ComponentTs', () => {
  beforeEach(() => {
    userState.reset()
  })

  it('renders correctly', () => {
    const wrapper = renderWithVuetify(ComponentTs)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders correctly with props', () => {
    const wrapper = renderWithVuetify(ComponentTs, {
      propsData: {
        text: 'My dummy Component',
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('given user name', () => {
    beforeEach(() => {
      const user = UserMother.create()
      userState.computedUserName.set(user.name)
    })

    it('renders correctly', () => {
      const wrapper = renderWithVuetify(ComponentTs)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders correctly with props', () => {
      const wrapper = renderWithVuetify(ComponentTs, {
        propsData: {
          text: 'My dummy Component',
        },
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('given random user name', () => {
    it('renders correctly', () => {
      const user = UserMother.random().build()
      userState.computedUserName.set(user.name)

      const wrapper = renderWithVuetify(ComponentTs)

      expect(wrapper.html()).toEqual(expect.stringContaining(user.name))
    })
  })
})
