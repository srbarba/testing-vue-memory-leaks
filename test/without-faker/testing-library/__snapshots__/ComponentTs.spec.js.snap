// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentTs given user name renders correctly 1`] = `
<div class="v-card v-sheet theme--light">Dummy component<div>Hello John Doe</div>
</div>
`;

exports[`ComponentTs given user name renders correctly with props 1`] = `
<div class="v-card v-sheet theme--light">My dummy Component<div>Hello John Doe</div>
</div>
`;

exports[`ComponentTs renders correctly 1`] = `
<div class="v-card v-sheet theme--light">Dummy component<div>Hello </div>
</div>
`;

exports[`ComponentTs renders correctly with props 1`] = `
<div class="v-card v-sheet theme--light">My dummy Component<div>Hello </div>
</div>
`;
